File: android/src/main/java/com/fromscratchstudio/peer2nodes/PeerCryptoAdapter.kt
package com.fromscratchstudio.peer2nodes

import java.util.Base64

/**
 * PeerCryptoAdapter — pure format-conversion utilities for P-256 public keys and
 * ECDSA signatures. Enables interoperability between platforms using different
 * serialisation conventions:
 *
 *   Platform        | Public key format           | Signature format
 *   ----------------|-----------------------------|------------------------------
 *   Node.js         | DER/SPKI  (91 bytes)        | DER ECDSA  (70–72 bytes)
 *   Android (JCE)   | DER/SPKI  (91 bytes)        | DER ECDSA  (70–72 bytes)
 *   iOS (CryptoKit) | x9.63 uncompressed (65 B)   | P1363 raw R‖S (64 bytes)
 *
 * Android (JCE) already uses canonical DER format natively. This adapter is provided
 * for completeness and for incoming messages from iOS peers.
 */
object PeerCryptoAdapter {

    // Fixed 26-byte DER/SPKI header for P-256 (secp256r1) public keys.
    private val P256_SPKI_HEADER = byteArrayOf(
        0x30.toByte(), 0x59.toByte(), 0x30.toByte(), 0x13.toByte(),
        0x06.toByte(), 0x07.toByte(), 0x2a.toByte(), 0x86.toByte(),
        0x48.toByte(), 0xce.toByte(), 0x3d.toByte(), 0x02.toByte(),
        0x01.toByte(), 0x06.toByte(), 0x08.toByte(), 0x2a.toByte(),
        0x86.toByte(), 0x48.toByte(), 0xce.toByte(), 0x3d.toByte(),
        0x03.toByte(), 0x01.toByte(), 0x07.toByte(), 0x03.toByte(),
        0x42.toByte(), 0x00.toByte()
    )

    // ── Public key conversions ────────────────────────────────────────────────

    /**
     * Converts an x9.63 uncompressed P-256 point (65 bytes, 0x04 prefix) to
     * DER/SPKI format (91 bytes).
     */
    fun x963ToDerSpki(x963Bytes: ByteArray): ByteArray {
        require(x963Bytes.size == 65 && x963Bytes[0] == 0x04.toByte()) {
            "x963ToDerSpki: expected 65-byte x9.63 uncompressed point, got ${x963Bytes.size} bytes"
        }
        return P256_SPKI_HEADER + x963Bytes
    }

    /**
     * Extracts the x9.63 uncompressed point (65 bytes) from a P-256 DER/SPKI key (91 bytes).
     */
    fun derSpkiToX963(derBytes: ByteArray): ByteArray {
        require(derBytes.size == 91) {
            "derSpkiToX963: expected 91-byte DER/SPKI, got ${derBytes.size} bytes"
        }
        require(derBytes.take(26).toByteArray().contentEquals(P256_SPKI_HEADER)) {
            "derSpkiToX963: DER/SPKI header does not match P-256 curve"
        }
        return derBytes.copyOfRange(26, 91)
    }

    // ── Signature conversions ─────────────────────────────────────────────────

    /**
     * Converts a P1363 signature (64-byte R‖S, CryptoKit native format) to
     * DER-encoded ECDSA (JCE / OpenSSL / Web Crypto format).
     */
    fun p1363ToDer(p1363Bytes: ByteArray): ByteArray {
        require(p1363Bytes.size == 64) {
            "p1363ToDer: expected 64-byte P1363 signature, got ${p1363Bytes.size} bytes"
        }
        val r    = derEncodeInteger(p1363Bytes.copyOfRange(0, 32))
        val s    = derEncodeInteger(p1363Bytes.copyOfRange(32, 64))
        val seq  = r + s
        return byteArrayOf(0x30.toByte(), seq.size.toByte()) + seq
    }

    /**
     * Converts a DER-encoded ECDSA signature to P1363 format (64-byte R‖S).
     */
    fun derToP1363(derBytes: ByteArray): ByteArray {
        require(derBytes.isNotEmpty() && derBytes[0] == 0x30.toByte()) {
            "derToP1363: expected DER SEQUENCE tag 0x30"
        }
        var pos = 2  // skip 0x30 and sequence-length byte
        val (r, rAdv) = decodeDerInteger(derBytes, pos); pos += rAdv
        val (s, _)    = decodeDerInteger(derBytes, pos)
        return padTo32(r) + padTo32(s)
    }

    // ── Auto-detecting normalisation ──────────────────────────────────────────

    /**
     * Normalises a base64-encoded P-256 public key to DER/SPKI format.
     * Accepts DER/SPKI (91 bytes) or x9.63 uncompressed (65 bytes, 0x04 prefix).
     * Returns base64-encoded DER/SPKI.
     */
    fun normalizePublicKey(base64Key: String): String {
        val bytes = Base64.getDecoder().decode(base64Key)
        if (bytes.size == 91 && bytes[0] == 0x30.toByte()) return base64Key
        if (bytes.size == 65 && bytes[0] == 0x04.toByte()) {
            return Base64.getEncoder().encodeToString(x963ToDerSpki(bytes))
        }
        error("normalizePublicKey: unrecognized P-256 key format (${bytes.size} bytes)")
    }

    /**
     * Normalises a base64-encoded P-256 ECDSA signature to DER format.
     * Accepts DER (starts with 0x30) or P1363 (exactly 64 bytes).
     * Returns base64-encoded DER signature.
     */
    fun normalizeSignature(base64Sig: String): String {
        val bytes = Base64.getDecoder().decode(base64Sig)
        if (bytes.isNotEmpty() && bytes[0] == 0x30.toByte()) return base64Sig
        if (bytes.size == 64) {
            return Base64.getEncoder().encodeToString(p1363ToDer(bytes))
        }
        error("normalizeSignature: unrecognized ECDSA signature format (${bytes.size} bytes)")
    }

    // ── Private helpers ───────────────────────────────────────────────────────

    /** DER INTEGER encoding: strip leading zeros, add 0x00 prefix if MSB is set. */
    private fun derEncodeInteger(bytes: ByteArray): ByteArray {
        var arr = bytes.toMutableList()
        while (arr.size > 1 && arr.first() == 0x00.toByte()) arr.removeFirst()
        if (arr.first().toInt() and 0x80 != 0) arr.add(0, 0x00.toByte())
        return byteArrayOf(0x02.toByte(), arr.size.toByte()) + arr.toByteArray()
    }

    /** Reads a DER INTEGER at bytes[pos]. Returns Pair(value, bytesConsumed). */
    private fun decodeDerInteger(bytes: ByteArray, pos: Int): Pair<ByteArray, Int> {
        require(pos < bytes.size && bytes[pos] == 0x02.toByte()) {
            "decodeDerInteger: expected INTEGER tag 0x02 at offset $pos"
        }
        val len = bytes[pos + 1].toInt() and 0xff
        return bytes.copyOfRange(pos + 2, pos + 2 + len) to (2 + len)
    }

    /** Strips leading 0x00 bytes and zero-pads on the left to exactly 32 bytes. */
    private fun padTo32(bytes: ByteArray): ByteArray {
        var arr = bytes.toMutableList()
        while (arr.size > 1 && arr.first() == 0x00.toByte()) arr.removeFirst()
        check(arr.size <= 32) { "padTo32: integer too long (${arr.size} bytes) for P-256" }
        while (arr.size < 32) arr.add(0, 0x00.toByte())
        return arr.toByteArray()
    }
}

File: ios/Sources/Peer2Nodes/PeerCryptoAdapter.swift
import Foundation

// MARK: - PeerCryptoAdapterError

public enum PeerCryptoAdapterError: Error, LocalizedError {
    case invalidFormat(String)

    public var errorDescription: String? {
        if case .invalidFormat(let msg) = self { return msg }
        return nil
    }
}

// MARK: - PeerCryptoAdapter
//
// Pure format-conversion utilities for P-256 public keys and ECDSA signatures.
// Enables interoperability between platforms using different serialisation conventions:
//
//   Platform        | Public key format           | Signature format
//   ----------------|-----------------------------|------------------------------
//   Node.js         | DER/SPKI  (91 bytes)        | DER ECDSA  (70–72 bytes)
//   Android (JCE)   | DER/SPKI  (91 bytes)        | DER ECDSA  (70–72 bytes)
//   iOS (CryptoKit) | x9.63 uncompressed (65 B)   | P1363 raw R‖S (64 bytes)
//
// All methods accept and return Data.

public enum PeerCryptoAdapter {

    // Fixed 26-byte DER/SPKI header for all P-256 (secp256r1) public keys.
    // Encodes: SEQUENCE { AlgorithmIdentifier { ecPublicKey OID, P-256 OID }, BIT STRING }
    private static let p256SpkiHeader = Data([
        0x30, 0x59, 0x30, 0x13, 0x06, 0x07, 0x2a, 0x86,
        0x48, 0xce, 0x3d, 0x02, 0x01, 0x06, 0x08, 0x2a,
        0x86, 0x48, 0xce, 0x3d, 0x03, 0x01, 0x07, 0x03,
        0x42, 0x00,
    ])

    // MARK: - Public key conversions

    /// Converts an x9.63 uncompressed P-256 point (65 bytes, 0x04 prefix) to DER/SPKI (91 bytes).
    public static func x963ToDerSpki(_ x963Data: Data) throws -> Data {
        guard x963Data.count == 65, x963Data[0] == 0x04 else {
            throw PeerCryptoAdapterError.invalidFormat(
                "x963ToDerSpki: expected 65-byte x9.63 uncompressed point, got \(x963Data.count) bytes"
            )
        }
        return p256SpkiHeader + x963Data
    }

    /// Extracts the x9.63 uncompressed point (65 bytes) from a DER/SPKI key (91 bytes).
    public static func derSpkiToX963(_ derData: Data) throws -> Data {
        guard derData.count == 91 else {
            throw PeerCryptoAdapterError.invalidFormat(
                "derSpkiToX963: expected 91-byte DER/SPKI, got \(derData.count) bytes"
            )
        }
        guard derData.prefix(26) == p256SpkiHeader else {
            throw PeerCryptoAdapterError.invalidFormat(
                "derSpkiToX963: DER/SPKI header does not match P-256 curve"
            )
        }
        return derData.dropFirst(26)
    }

    // MARK: - Signature conversions

    /// Converts a P1363 signature (64-byte R‖S, CryptoKit native) to DER-encoded ECDSA.
    public static func p1363ToDer(_ p1363Data: Data) throws -> Data {
        guard p1363Data.count == 64 else {
            throw PeerCryptoAdapterError.invalidFormat(
                "p1363ToDer: expected 64-byte P1363 signature, got \(p1363Data.count) bytes"
            )
        }
        let r      = derEncodeInteger(p1363Data.prefix(32))
        let s      = derEncodeInteger(p1363Data.suffix(32))
        let seq    = r + s
        return Data([0x30, UInt8(seq.count)]) + seq
    }

    /// Converts a DER-encoded ECDSA signature to P1363 format (64-byte R‖S).
    public static func derToP1363(_ derData: Data) throws -> Data {
        guard derData.first == 0x30 else {
            throw PeerCryptoAdapterError.invalidFormat("derToP1363: expected DER SEQUENCE tag 0x30")
        }
        var pos = 2   // skip 0x30 and sequence-length byte
        let (r, rAdv) = try decodeDerInteger(derData, at: pos); pos += rAdv
        let (s, _)    = try decodeDerInteger(derData, at: pos)
        return padTo32(r) + padTo32(s)
    }

    // MARK: - Auto-detecting normalisation

    /// Normalises a P-256 public key to DER/SPKI format.
    /// Accepts DER/SPKI (91 bytes) or x9.63 uncompressed (65 bytes, 0x04 prefix).
    public static func normalizePublicKey(_ data: Data) throws -> Data {
        if data.count == 91 && data[0] == 0x30 { return data }
        if data.count == 65 && data[0] == 0x04 { return try x963ToDerSpki(data) }
        throw PeerCryptoAdapterError.invalidFormat(
            "normalizePublicKey: unrecognized P-256 key format (\(data.count) bytes)"
        )
    }

    /// Normalises a P-256 ECDSA signature to DER format.
    /// Accepts DER (starts with 0x30) or P1363 (exactly 64 bytes).
    public static func normalizeSignature(_ data: Data) throws -> Data {
        if data.first == 0x30     { return data }
        if data.count == 64       { return try p1363ToDer(data) }
        throw PeerCryptoAdapterError.invalidFormat(
            "normalizeSignature: unrecognized ECDSA signature format (\(data.count) bytes)"
        )
    }

    // MARK: - Private helpers

    /// DER INTEGER encoding: strip leading zero bytes, add 0x00 prefix if MSB is set.
    private static func derEncodeInteger(_ bytes: Data) -> Data {
        var arr = Array(bytes)
        while arr.count > 1 && arr.first == 0x00 { arr.removeFirst() }
        if arr.first! & 0x80 != 0 { arr.insert(0x00, at: 0) }
        return Data([0x02, UInt8(arr.count)]) + Data(arr)
    }

    /// Reads a DER INTEGER at `data[pos]`. Returns (value, bytesConsumed).
    private static func decodeDerInteger(_ data: Data, at pos: Int) throws -> (Data, Int) {
        guard pos < data.count, data[pos] == 0x02 else {
            throw PeerCryptoAdapterError.invalidFormat(
                "decodeDerInteger: expected INTEGER tag 0x02 at offset \(pos)"
            )
        }
        let len = Int(data[pos + 1])
        let val = data[(pos + 2)..<(pos + 2 + len)]
        return (val, 2 + len)
    }

    /// Strips leading 0x00 bytes and zero-pads on the left to exactly 32 bytes.
    private static func padTo32(_ data: Data) -> Data {
        var stripped = data
        while stripped.first == 0x00 { stripped = stripped.dropFirst() }
        guard stripped.count <= 32 else {
            return Data(stripped.suffix(32)) // should not happen for valid P-256 values
        }
        let padding = Data(repeating: 0, count: 32 - stripped.count)
        return padding + stripped
    }
}
